src/VisualStudio/Core/Impl/SolutionExplorer/SourceGeneratedFileItems/SourceGeneratedFileItemSource.cs

```diff
@@ -0,0 +1,281 @@
+                                high = mid;
+                            }
+                            else
+                            {
+                                low = mid + 1;
+                            }
+                        }
+
+                        _items.Insert(low, new SourceGeneratedFileItem(document.Id, document.HintName, document.Project.Language, _workspace));
+                    }
+                }
+                else
+                {
+                    // We don't have any items at all, so add the placeholder
+                    Contract.ThrowIfFalse(_items.Count == 0);
+                    _items.Add(new NoSourceGeneratedFilesPlaceholderItem());
+                }
+            }
+            finally
+            {
+                _items.EndBulkOperation();
+                _items.MarkAsInitialized();
+            }
+        }
+
+        public void BeforeExpand()
+        {
+            lock (_gate)
+            {
+                // We should not have an existing computation active
+                Contract.ThrowIfNull(_cancellationTokenSource == null);
+
+                _cancellationTokenSource = new CancellationTokenSource();
+                var cancellationToken = _cancellationTokenSource.Token;
+                var asyncToken = _asyncListener.BeginAsyncOperation(nameof(SourceGeneratedFileItemSource) + "." + nameof(BeforeExpand));
+
+                Task.Run(
+                    async () =>
+                    {
+                        // Since the user just expanded this, we want to do a single population aggressively,
+                        // where the only reason we'd cancel is if the user collapsed it again.
+                        var solution = _workspace.CurrentSolution;
+                        await UpdateSourceGeneratedFileItemsAsync(solution, cancellationToken).ConfigureAwait(false);
+
+                        // Now that we've done it the first time, we'll subscribe for future changes
+                        lock (_gate)
+                        {
+                            // It's important we check for cancellation inside our lock: if the user were to collapse
+                            // right at this point, we don't want to have a case where we cancelled the work, unsubscribed
+                            // in AfterCollapse, and _then_ subscribed here again.
+
+                            cancellationToken.ThrowIfCancellationRequested();
+                            _workspace.WorkspaceChanged += OnWorkpaceChanged;
+                            if (_workspace.CurrentSolution != solution)
+                            {
+                                // The workspace changed while we were doing our initial population, so
+                                // refresh it. We'll just call our OnWorkspaceChanged event handler
+                                // so this looks like any other change.
+                                OnWorkpaceChanged(this,
```

Comment:
would it be simpler at all if when this node was expanded it just triggered a poll every 5 seconds (maybe chained task?) that would check if the workspace changed since the last check and update the nodes if it had?  Then you don't need to deal with the workspace change subscription / resettable delay.

I guess you'd be wasting a little bit of processing if VS is just idle (workspace isn't changing) but it doesn't seem like much since it's very infrequent

Reply:
My worry was I didn't want to start doing eager pulling on source generators during active user typing. In some cases it wouldn't matter if you have the nodes expanded in a project you're editing, but imagine a case where you are editing in A, and B, depends on A, and C depends on B...getting the updated generators for C isn't cheap. If you don't have any files open from C we won't be actively creating compilations for C, but having this expanded now would.

I agree this code isn't as simple as I'd like but I'm not sure I want to go to a simple timer.